src/transaction-flow/input/SelectPrimaryName/SelectPrimaryName-flow.tsx

```diff
@@ -0,0 +1,381 @@
+  `,
+  mq.sm.min(css`
+    width: calc(100% + 2 * ${theme.space['6']});
+    margin: 0 -${theme.space['6']};
+  `),
+])
+
+const NameTableHeaderWrapper = styled.div(({ theme }) => [
+  css`
+    margin: 0 -${theme.space['4']};
+    > div {
+      border-bottom: none;
+    }
+  `,
+  mq.sm.min(css`
+    margin: 0 -${theme.space['4.5']};
+  `),
+])
+
+const StyledScrollBox = styled(ScrollBoxWithSpinner)(
+  ({ theme }) => css`
+    width: ${theme.space.full};
+
+    & > div:nth-last-child(2) {
+      border-bottom: none;
+    }
+  `,
+)
+
+const ErrorContainer = styled.div(
+  ({ theme }) => css`
+    display: flex;
+    align-items: center;
+    justify-content: center;
+    padding: ${theme.space['4']};
+  `,
+)
+
+const SelectPrimaryName = ({ data: { address }, dispatch, onDismiss }: Props) => {
+  const { t } = useTranslation('transactionFlow')
+  const formRef = useRef<HTMLFormElement>(null)
+  const queryClient = useQueryClient()
+
+  const form = useForm<FormData>({
+    mode: 'onChange',
+    defaultValues: {
+      name: undefined,
+      unknownLabels: {
+        tld: '',
+        labels: [],
+      },
+    },
+  })
+  const { handleSubmit, control, setValue } = form
+
+  const provider = useProvider()
+  const chainId = useChainId()
+  const lastestResolverAddress = RESOLVER_ADDRESSES[`${chainId}`]?.[0]
+
+  const { ready: isEnsReady, getResolver } = useEns()
+
+  const [view, setView] = useState<'main' | 'decrypt'>('main')
+
+  const [sortType, setSortType] = useState<SortType>('labelName')
+  const [sortDirection, setSortDirection] = useState<SortDirection>('asc')
+  const [searchInput, setSearchInput] = useState('')
+  const [searchQuery, _setSearchQuery] = useState('')
+  const setSearchQuery = useDebouncedCallback(_setSearchQuery, 300, [])
+
+  const {
+    names: currentPage,
+    count: namesCount,
+    loadMore: loadMoreNames,
+    isLoading: isLoadingNames,
+  } = useAvailablePrimaryNamesForAddress({
+    address,
+    sort: {
+      type: sortType,
+      orderDirection: sortDirection,
+    },
+    search: searchQuery,
+    resultsPerPage: DEFAULT_PAGE_SIZE,
+  })
+
+  const isLoading = !isEnsReady || isLoadingNames
+
+  const selectedName = useWatch({
+    control,
+    name: 'name.name',
+  })
+
+  const onSubmit = async (data: FormData) => {
+    if (!data.name || isLoading) return
+
+    let validName = data.name.name
+
+    const hasEncodedLabel = validName.split('.').some((label) => isEncodedLabelhash(label))
+    const hasValidDecodedLabels =
+      data.unknownLabels.labels.length === validName.split('.').length - 1 &&
+      data.unknownLabels.labels.every(
+        ({ label, value }) => label === value || label === labelhash(value),
+      )
+
+    // If name has encoded labels and we do not have the decoded labels, switch to decrypt view
+    if (hasEncodedLabel && !hasValidDecodedLabels) {
+      setValue('unknownLabels', nameToFormData(data.name.name).unknownLabels)
+      setView('decrypt')
+      return
+    }
```

Comment:
is it worth trying to get the decoded label with `getDecryptedName()` from ensjs?